avm/res/sql/server/elastic-pool/main.bicep

```
@@ -68,7 +68,7 @@ resource elasticPool 'Microsoft.Sql/servers/elasticPools@2023-08-01-preview' = {
   properties: {
     highAvailabilityReplicaCount: highAvailabilityReplicaCount
     licenseType: licenseType
-    maintenanceConfigurationId: maintenanceConfigurationId
+    maintenanceConfigurationId: !empty(maintenanceConfigurationId) ? maintenanceConfigurationId : null
```

```suggestion
    maintenanceConfigurationId: maintenanceConfigurationResourceId
```
Should contain `resourceId`. Also, I'd recommend to make the parameter nullable through which you don't need this logic as the default would be `null` anyways

---

@AlexanderSehr I have done further tests with this, and based on the tests this empty check IS necessary.

I have tested the following options:
* `param maintenanceConfigurationId string = ''`
* `param maintenanceConfigurationId string?` 

If there is no empty check (just passing on the incoming parameter), then both of these lead to the same error:

`Property id '' at path 'properties.maintenanceConfigurationId' is invalid. Expect fully qualified resource Id that
     | start with '/subscriptions/{subscriptionId}' or '/providers/{resourceProviderNamespace}/'.`

The only way to specify a deployment without a maintenance window is to check for empty parameter and pass the `null` explicitly:

`maintenanceConfigurationId: !empty(maintenanceConfigurationId) ? maintenanceConfigurationId : null`

So based on this I'd like to put this back.

---

Hey @peterbud,
it should depend on how you want the user experience to be like. To explain myself:
- If you want you user to be able to specify an empty input value for an optional parameter you do need the input check, yes. In that case you can also have that default value `''` because your empty check will detect that no value was provided and pass `null` into to the RP on behalf of the user (this is the currently implemented behavior)
  ```bicep
  param aParameter string = ''
  
  resource dummy '(...)' = {
    properties: {
      aProperty: !empty(aParameter) ? aParameter : null
    }
  }
  ```
- The one I usually prefer (but it's absolutely up to you) is the one where I want the user to either specify a value, not specify anything at all, or if neither is an option, specify `null`. In this case, the parameter would be implemented as nullable, and the place where the parameter is used would not need an empty check, as the value allowed values are anyways only either `null` or `a proper value`.
  ```bicep
  param aParameter string?
  
  resource dummy '(...)' = {
    properties: {
      aProperty: aParameter
    }
  }
  ```

The only real difference of the two is that the user cannot specify `''` because this value would be passed through to the RP which sometimes does not support it.

Last but not least, I suspect the error you're seeing is actually not from this implementation but the way the parameter is passed in from the parent module: `maintenanceConfigurationId: elasticPool.?maintenanceConfigurationId ?? ''`. That that it really really wants to pass in `''` if the user did not specify the parameter. This should be removed and it should just say `maintenanceConfigurationId: elasticPool.?maintenanceConfigurationId` which will simply pass in `null` to the child module (or, if it would have a default there, would use that).
That last point is by the way something I've been slowly but surely cleaning up in the AVM modules I worked on. As pass null to a child means you assume the corresponding parameter's default value, the default in the parent I just called out is not needed anymore. This was only added because back in the days it was technically required to provide a default to make a parameter optional. Hence my advice: For all parameters that are passed to any other bicep file, check if the default the parent may specify via the `??` is the same as the default in the child. If so, you can safely remove the `?? (...)`  content in the parent.

> Sidenote (just some trivia): `empty(null)` evaluates to `true`

---

Long story short, both implementations will work. Just wanted to give a recommendation which should work if you remove the default from the parent 😉

---

Thanks, that has really helped I think. I'd also prefer the second, and I was wondering why actually has failed. Now I get it. Let me fix this.